README.md
Crashwalk
=======

## Documentation

If you want to use `import "github.com/bnagy/crashwalk"` in your own Go code, you
can get godoc at: http://godoc.org/github.com/bnagy/crashwalk

## cwtriage

To run the standalone cwtriage tool:
```
  cwtriage runs crashfiles with instrumentation and outputs results in various formats
  Usage: cwtriage -root /path/to/afl-dir [-match pattern] -- /path/to/target -in @@ -out whatever
  ( @@ will be substituted for each crashfile )

  -afl
      Prefer the AFL recorded crashing command, if present
  -engine string
      Debugging engine to use: [gdb lldb] (default "gdb")
  -every int
      Run every n seconds (default -1)
  -f string
      Template filename to use while running crash
  -ignore string
      Directory skip pattern ( go regex syntax )
  -match string
      Match pattern for files ( go regex syntax )
  -mem int
      Memory limit for target processes (MB) (default -1)
  -output string
      Output format to use: [json pb text] (default "text")
  -root string
      Root directory to look for crashes
  -seen
      Include seen results from the DB in the output
  -strict
      Abort the whole run if any crashes fail to repro
  -t int
      Timeout for target processes (secs) (default 60)
  -tidy
      Move crashes that error under Run() to a tidy dir
  -workers int
```

### AFL Mode

If you're using AFL >= 1.50b then afl automatically records the command that was used in each crash dir ( in the README.txt file ). For most people, that means you can use the -afl switch to:
- Automatically set the -match pattern to match AFL crashfiles
- Automatically use the stored command from the README.txt in each crash directory
- Automatically set the same memory limit
- Automatically ignore queue/ and hang/ directories
- Automatically pick up `-f` template configuration from README.txt
- Use the supplied command (if any) as a default

If you are in the directory that contains your individual fuzz workers, then the minimal command would be something like
```
cwtriage -root . -afl
```

**NOTE** See [Bugs](#bugs) below!

### Manual Mode

If you're triaging from an older version of AFL or you want to run the crashes with a different target command, then `-match crashes.\*id` should match AFL crashes.

The tool creates a BoltDB (in the current directory, by default) that is used to cache crash instrumentation results. This way you can run it multiple times on an active directory and only get the latest crashes, or run it with -seen and get all crashes, but faster. To "reset the cache" just `rm crashwalk.db`. NOTE that the cache DOES NOT contain the actual crashfile, only the metadata, so don't go deleting your crashes or anything.

### Output Formats

Supported output formats are JSON, protocol buffers or the text summary seen in the examples below. JSON and protbuf output is one crash per line, to facilitate piping that output to another process - for example to push each crash to a queue, write them to a database etc.

### Tidy

When you have crashfiles that don't repro under the debugger they are not added to the `cwtriage` cache database, which means that `cwtriage` will attempt to run then every time, even without `-seen`. If you have a lot of these files, or if they're particularly slow (such a memory eaters and hangs) then they can add a lot of time to each run. By using the `-tidy` flag, files that don't crash under the debugger will be moved to a directory called `.cwtidy` inside the crash directory.

### -f

If you have an app that expects a certain extension you can use the `-f` option, with some limitations. Because we support multiple workers, you can't specify an exact output file. Any file you specify, like `/dev/shm/blah.txt` will be used as a template, and each worker will copy the crashdata for each crash into a randomised 8 character name like `/dev/shm/fjsyvnsh.txt`, cleaning up at the end.

*NOTE* `-afl` mode will automatically do this for you if you used a `-f` option to `afl-fuzz` or `afl-launch`, so you don't need to pass this option.

## cwdump

`cwdump` summarizes crashes in a crashwalk database by major / minor stack hash. Although AFL (for example) already de-dupes crashes, bucketing summarizes those crashes by an order of magnitude or more. Crashes that bucket the same have _exactly_ the same stack contents, so they're likely (not guaranteed) to be the same bug.

Run `cwtriage` as above and then do something like
```
cwdump ./crashwalk.db > triage.txt
```

## cwfind

`cwfind` is a simple utility to output the filenames of all crashes matching a given hash. I use it in combination with `xargs` to bulk delete / move crashfiles.

```
$ cwfind a9060880abffbe2dcd5c9b4bb39c9233.0e358881a2545b216eee9aabe3723302
pdf-S3/crashes/id:000821,sig:08,src:019017+009441,op:splice,rep:16
pdf-S25/crashes/id:000823,sig:08,src:019101+013412,op:splice,rep:64
pdf-S17/crashes/id:000870,sig:08,src:025983+017873,op:splice,rep:8
pdf-S3/crashes/id:000822,sig:08,src:019017+009441,op:splice,rep:4
```
## Installation

1. You should follow the [instructions](https://golang.org/doc/install) to
install Go, if you haven't already done so.

2. (FOR LINUX ONLY) Install the 'exploitable' tool from [here](https://github.com/jfoote/exploitable). Right now the code is expecting to find the tool at `~/src/exploitable/exploitable/exploitable.py`. If this is impossible, you can set the CW_EXPLOITABLE environment variable and it should get picked up.

3. (FOR LINUX ONLY) Make sure you have gdb in your path with `which gdb`

4. (FOR OSX ONLY) I wrote a very heavily modified mutant offspring of exploitable and one of the lldb sample tools, called `exploitaben.py`. Unless you do something unusual the `cwtriage` binary will install it as a dependency and use it for `-engine lldb`. You can check the lldb specific code [here](https://github.com/bnagy/francis)

5. (FOR OSX ONLY) Make sure lldb is installed. You might need to mess about with assorted Xcode hijinks etc.

Now, install crashwalk:
```bash
$ go get -u github.com/bnagy/crashwalk/cmd/...
```

The binaries produced are statically linked (Go just does that), so you can 'deploy' to other systems, docker containers etc by just copying them.

No overarching tests yet, sorry, it's a little fiddly to build a standalone testbed. The gdb / lldb parsers will panic if they get confused and give you the problematic input and a useful stack trace. If the input is not sensitive, use that to open an issue and I'll fix it.

`exploitaben` does have a reasonable set of tests, and you can check out [the results](https://github.com/bnagy/francis/exploitaben/test_results/test_output.txt). I suck at python so the diffing is not automatic. I included all of the exploitable project tests as well as all the CrashWrangler tests.

## Screenshots

#### GDB Example - unique faulting EIPs
```
./cwtriage -afl -root . -workers 16 | grep =\> | sort | uniq -c
2015/02/19 00:58:20 Workers started
2015/02/19 01:06:51 All done!
      1 => 0x00007ffff6184dfe: push r13
    140 => 0x00007ffff6184e17: mov DWORD PTR [rbp-0x4e0],eax
     75 => 0x00007ffff6184e7d: call 0x7ffff61cf5e0 <strchrnul>
      3 => 0x00007ffff6184fa9: call 0x7ffff6189f90 <buffered_vfprintf>
      1 => 0x00007ffff6189f92: push r12
    129 => 0x00007ffff6189fc5: mov QWORD PTR [rsp+0x100],rbx
     12 => 0x00007ffff6189fd8: mov DWORD PTR [rsp+0x20],0xfbad8004
      1 => 0x00007ffff61b70da: push r13
      5 => 0x00007ffff61b70de: push rbp
      1 => 0x00007ffff61b71c4: call 0x7ffff61c8230 <__mempcpy_sse2>
```

#### GDB Example - summary output
```
---CRASH SUMMARY---
Filename: /dev/shm/crashexplore/fuzz3/crashes/id:000359,sig:11,src:001295,op:havoc,rep:16
SHA1: 30d6e81570a67cb32bbd51b460d74bb193a85d98
Classification: EXPLOITABLE
Hash: 2f50f8e5a6fb7a55669dc8ead34fdba3.ba6f36a1a8447da016a175a99706a64b
Command: /home/ben/src/poppler-0.26.5/utils/pdftocairo -jpeg /dev/shm/crashexplore/fuzz3/crashes/id:000359,sig:11,src:001295,op:havoc,rep:16
Faulting Frame:
   fprintf @ 0x000000000044f7be: in /home/ben/src/poppler-0.26.5/utils/pdftocairo
Disassembly:
   0x00007ffff6184e60: mov QWORD PTR [rbp-0x460],rax
   0x00007ffff6184e67: mov rax,QWORD PTR [r15+0x8]
   0x00007ffff6184e6b: mov QWORD PTR [rbp-0x458],rax
   0x00007ffff6184e72: mov rax,QWORD PTR [r15+0x10]
   0x00007ffff6184e76: mov QWORD PTR [rbp-0x450],rax
=> 0x00007ffff6184e7d: call 0x7ffff61cf5e0 <strchrnul>
   0x00007ffff6184e82: and r12d,0x8000
   0x00007ffff6184e89: mov QWORD PTR [rbp-0x4e8],rax
   0x00007ffff6184e90: mov QWORD PTR [rbp-0x4d0],rax
   0x00007ffff6184e97: mov DWORD PTR [rbp-0x4dc],0x0
Stack Head (1001 entries):
   _IO_vfprintf_internal     @ 0x00007ffff6184e7d: in /lib/x86_64-linux-gnu/libc-2.19.so (BL)
   buffered_vfprintf         @ 0x00007ffff618a021: in /lib/x86_64-linux-gnu/libc-2.19.so (BL)
   _IO_vfprintf_internal     @ 0x00007ffff6184fae: in /lib/x86_64-linux-gnu/libc-2.19.so (BL)
   ___fprintf_chk            @ 0x00007ffff6245065: in /lib/x86_64-linux-gnu/libc-2.19.so (BL)
   fprintf                   @ 0x000000000044f7be: in /home/ben/src/poppler-0.26.5/utils/pdftocairo
   error                     @ 0x000000000044f7be: in /home/ben/src/poppler-0.26.5/utils/pdftocairo
   Parser::getObj            @ 0x0000000000569363: in /home/ben/src/poppler-0.26.5/utils/pdftocairo
   Parser::getObj            @ 0x0000000000568deb: in /home/ben/src/poppler-0.26.5/utils/pdftocairo
   Parser::getObj            @ 0x000000000056997b: in /home/ben/src/poppler-0.26.5/utils/pdftocairo
   XRef::fetch               @ 0x00000000005d2d01: in /home/ben/src/poppler-0.26.5/utils/pdftocairo
   dictLookup                @ 0x00000000005aa3a5: in /home/ben/src/poppler-0.26.5/utils/pdftocairo
   Stream::addFilters        @ 0x00000000005aa3a5: in /home/ben/src/poppler-0.26.5/utils/pdftocairo
   Parser::makeStream        @ 0x000000000056842e: in /home/ben/src/poppler-0.26.5/utils/pdftocairo
   Parser::getObj            @ 0x00000000005695d2: in /home/ben/src/poppler-0.26.5/utils/pdftocairo
   Parser::getObj            @ 0x000000000056997b: in /home/ben/src/poppler-0.26.5/utils/pdftocairo
   Parser::getObj            @ 0x000000000056997b: in /home/ben/src/poppler-0.26.5/utils/pdftocairo
Registers:
rax=0x00007fffff801ca0 rbx=0x00007fffff7ff560 rcx=0x00000000ffffffff rdx=0x00007fffff801c88 
rsi=0x0000000000000025 rdi=0x000000000083976a rbp=0x00007fffff7ff530 rsp=0x00007fffff7fef50 
 r8=0x00007ffff64fb9f0  r9=0x0000000001f21e10 r10=0x000000000083976a r11=0x0000000000aa8000 
r12=0x00000000fbad8004 r13=0x0000000000000024 r14=0x000000000083976a r15=0x00007fffff801c88 
rip=0x00007ffff6184e7d efl=0x0000000000010246  cs=0x0000000000000033  ss=0x000000000000002b 
 ds=0x0000000000000000  es=0x0000000000000000  fs=0x0000000000000000  gs=0x0000000000000000 
Extra Data:
   Description: Access violation during branch instruction
   Short description: BranchAv (4/22)
   Explanation: The target crashed on a branch instruction, which may indicate that the control flow is tainted.
---END SUMMARY---
```

### LLDB Example - summary output

For `exploitaben` I used "indicators" instead of a single classification. I'm hoping that this is more flexible and better suited to post-crash database searches and such. They're in the `Extra Data` field below, so that the crash summary protobufs are crossplatform.
```
---CRASH SUMMARY---
Filename: /Volumes/ramdisk/popplUNFIXED/crashes/id:000156,sig:06,src:006567,op:havoc,rep:16
SHA1: 53100e26a2eced52afa5a8f1708ced7a6dbb6435
Classification: 
Hash: 8263c96a28b14639c338ffd408d3701e.0add20a3424f9105c8476367f36eb255
Command: /Users/ben/src/poppler-0.31.0/utils/pdftoppm -aa no -r 36 -png /Volumes/ramdisk/popplUNFIXED/crashes/id:000156,sig:06,src:006567,op:havoc,rep:16
Faulting Frame:
   Catalog::getNumPages() [inlined] Object::dictIs(this=0x000a000000000007, dictType=<unavailable>) - 18446744069413943243 @ 0x000000010009c834: in pdftoppm
Disassembly:

CHANGES
Apr 24, 2015:

- Add json and protobuf output options to afl-triage. With the new options,
  each crash is emitted as a single line, marshalled as either JSON or a
  protobuf "compact string", suitable for piping to another process.

Jun 17, 2015:

- Add MemoryLimit and Timeout options to CrashwalkConfig struct
- Add corresponding -t and -mem options to afl-triage
- Change signature of Debugger interface ( internal / API change)
- -auto flag now automatically sets memory limit from afl's README.txt
- Fix some stupid hang bugs in the gdb engine

Jun 26, 2015:

- Fix a parsing bug in the gdb engine for the "faulting frame" struct

Jun 28, 2015:

- Add cwfind and cwdump commands
- Rename afl-triage to cwtriage

Jul 15, 2015:

- Rename -auto option to -afl
- Add -ignore option to cwtriage to skip whole dirs

August 13, 2015:

- cwtriage - add support for triaging targets that take input over stdin
- this fix required breaking the Debugger interface, so francis also needs
  an update.

August 28, 2015:

- cwtriage - substitute @@ markers inside argument strings (Issue #5)
- cwtriage - add -tidy flag to move crashes that don't repro in cwtidy directories

September 2, 2015:

- cwtriage - change the location of tidied crashes. Now in .cwtidy inside the crash dir.

September 6, 2015:

- Add limited -f support, see README.md

November 5, 2015:

- -afl now picks up -m and -f options from README.txt
- (internal) crashwalk now dispatches job options per job instead of via global config

January 20, 2015:

- the GDB exploitable plugin can now be located using the CW_EXPLOITABLE environment variable (PR #10 zuBux)

lldb/lldb.go
package lldb

// This is a parsing wrapper for LLDB. It is designed for use with this tool,
// but may have independent utility, in which case I'll refactor it out later.
// It implements the simple Debugger() interface required by the crashwalk
// package. Essentially, crashwalk takes care of finding the files, this code
// takes care of running them under LLDB and parsing the output into
// crashwalk/crash's struct formats

import (
	"bufio"
	"bytes"
	"fmt"
	"github.com/bnagy/crashwalk/crash"
	"go/build"
	"io/ioutil"
	"os/exec"
	"path/filepath"
	"regexp"
	"strconv"
	"strings"
)

type Engine struct{}

// Frames in this blacklist are not eligible to be considered the faulting
// frame.
var libsystemRgx = regexp.MustCompile("libsystem")
var blacklist = []*regexp.Regexp{libsystemRgx}

// Example output
/*
Stack trace:
* thread #1: tid = 0x154da494, 0x00007fff97a33c9c libsystem_c.dylib`__vfprintf + 145, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=2, address=0x7fff5f3fffe8)
  * frame #0: 0x00007fff97a33c9c libsystem_c.dylib`__vfprintf + 145
    frame #1: 0x00007fff97a5d9cf libsystem_c.dylib`__v2printf + 679
    frame #2: 0x00007fff97a5dc64 libsystem_c.dylib`__xvprintf + 536
    frame #3: 0x00007fff97a33b92 libsystem_c.dylib`vfprintf_l + 54
    frame #4: 0x00007fff97a2c620 libsystem_c.dylib`fprintf + 186
    frame #5: 0x00000001000ca176 pdftoppm`error(category=<unavailable>, pos=<unavailable>, msg=<unavailable>) - 18446744069413756553
    frame #6: 0x00000001002783dd pdftoppm`Parser::makeStream(this=<unavailable>, dict=<unavailable>, fileKey=<unavailable>, encAlgorithm=<unavailable>, keyLength=<unavailable>, objNum=<unavailable>, objGen=<unavailable>, recursion=<unavailable>, strict=<unavailable>) - 18446744069411994658
    frame #7: 0x00000001002777d7 pdftoppm`Parser::getObj(this=<unavailable>, obj=<unavailable>, simpleOnly=<unavailable>, fileKey=<unavailable>, encAlgorithm=<unavailable>, keyLength=<unavailable>, objNum=<unavailable>, objGen=<unavailable>, recursion=<unavailable>, strict=<unavailable>) - 18446744069411997736
    frame #8: 0x000000010027685e pdftoppm`Parser::getObj(this=<unavailable>, obj=<unavailable>, simpleOnly=<unavailable>, fileKey=<unavailable>, encAlgorithm=<unavailable>, keyLength=<unavailable>, objNum=<unavailable>, objGen=<unavailable>, recursion=<unavailable>, strict=<unavailable>) - 18446744069412001697
    frame #9: 0x00000001002fefbc pdftoppm`XRef::fetch(this=<unavailable>, num=<unavailable>, gen=<unavailable>, obj=<unavailable>, recursion=<unavailable>) - 18446744069411442755
    frame #10: 0x00000001000c48d3 pdftoppm`Dict::lookup(this=<unavailable>, key=<unavailable>, obj=<unavailable>, recursion=<unavailable>) - 18446744069413779244
    frame #11: 0x000000010029b28e pdftoppm`Stream::addFilters(Object*, int) [inlined] Object::dictLookup(this=<unavailable>, key=<unavailable>, obj=0x0000000100000007, recursion=<unavailable>) - 18446744069411851633
    frame #12: 0x0000000100278fd6 pdftoppm`Parser::makeStream(this=<unavailable>, dict=<unavailable>, fileKey=<unavailable>, encAlgorithm=<unavailable>, keyLength=<unavailable>, objNum=<unavailable>, objGen=<unavailable>, recursion=<unavailable>, strict=<unavailable>) - 18446744069411991593
    frame #13: 0x00000001002777d7 pdftoppm`Parser::getObj(this=<unavailable>, obj=<unavailable>, simpleOnly=<unavailable>, fileKey=<unavailable>, encAlgorithm=<unavailable>, keyLength=<unavailable>, objNum=<unavailable>, objGen=<unavailable>, recursion=<unavailable>, strict=<unavailable>) - 18446744069411997736
    frame #14: 0x000000010027685e pdftoppm`Parser::getObj(this=<unavailable>, obj=<unavailable>, simpleOnly=<unavailable>, fileKey=<unavailable>, encAlgorithm=<unavailable>, keyLength=<unavailable>, objNum=<unavailable>, objGen=<unavailable>, recursion=<unavailable>, strict=<unavailable>) - 18446744069412001697
    frame #15: 0x00000001002fefbc pdftoppm`XRef::fetch(this=<unavailable>, num=<unavailable>, gen=<unavailable>, obj=<unavailable>, recursion=<unavailable>) - 18446744069411442755
    frame #16: 0x00000001000c48d3 pdftoppm`Dict::lookup(this=<unavailable>, key=<unavailable>, obj=<unavailable>, recursion=<unavailable>) - 18446744069413779244
    frame #17: 0x000000010029b28e pdftoppm`Stream::addFilters(Object*, int) [inlined] Object::dictLookup(this=<unavailable>, key=<unavailable>, obj=0x0000000100000007, recursion=<unavailable>) - 18446744069411851633
    frame #18: 0x0000000100278fd6 pdftoppm`Parser::makeStream(this=<unavailable>, dict=<unavailable>, fileKey=<unavailable>, encAlgorithm=<unavailable>, keyLength=<unavailable>, objNum=<unavailable>, objGen=<unavailable>, recursion=<unavailable>, strict=<unavailable>) - 18446744069411991593
    frame #19: 0x00000001002777d7 pdftoppm`Parser::getObj(this=<unavailable>, obj=<unavailable>, simpleOnly=<unavailable>, fileKey=<unavailable>, encAlgorithm=<unavailable>, keyLength=<unavailable>, objNum=<unavailable>, objGen=<unavailable>, recursion=<unavailable>, strict=<unavailable>) - 18446744069411997736
    frame #20: 0x000000010027685e pdftoppm`Parser::getObj(this=<unavailable>, obj=<unavailable>, simpleOnly=<unavailable>, fileKey=<unavailable>, encAlgorithm=<unavailable>, keyLength=<unavailable>, objNum=<unavailable>, objGen=<unavailable>, recursion=<unavailable>, strict=<unavailable>) - 18446744069412001697
    frame #21: 0x00000001002fefbc pdftoppm`XRef::fetch(this=<unavailable>, num=<unavailable>, gen=<unavailable>, obj=<unavailable>, recursion=<unavailable>) - 18446744069411442755
    frame #22: 0x00000001000c48d3 pdftoppm`Dict::lookup(this=<unavailable>, key=<unavailable>, obj=<unavailable>, recursion=<unavailable>) - 18446744069413779244
    frame #23: 0x000000010029b28e pdftoppm`Stream::addFilters(Object*, int) [inlined] Object::dictLookup(this=<unavailable>, key=<unavailable>, obj=0x0000000100000007, recursion=<unavailable>) - 18446744069411851633
    frame #24: 0x0000000100278fd6 pdftoppm`Parser::makeStream(this=<unavailable>, dict=<unavailable>, fileKey=<unavailable>, encAlgorithm=<unavailable>, keyLength=<unavailable>, objNum=<unavailable>, objGen=<unavailable>, recursion=<unavailable>, strict=<unavailable>) - 18446744069411991593

Nearby code:
libsystem_c.dylib`__vfprintf + 145:
-> 0x7fff97a33c9c:  e8 d9 01 04 00        call   0x7fff97a73e7a            ; symbol stub for: localeconv_l
   0x7fff97a33ca1:  48 8b 18              mov    rbx, qword ptr [rax]
   0x7fff97a33ca4:  80 7b 01 00           cmp    byte ptr [rbx + 0x1], 0x0
   0x7fff97a33ca8:  b8 01 00 00 00        mov    eax, 0x1
   0x7fff97a33cad:  74 08                 je     0x7fff97a33cb7            ; __vfprintf + 172
   0x7fff97a33caf:  48 89 df              mov    rdi, rbx
   0x7fff97a33cb2:  e8 a7 fc 03 00        call   0x7fff97a7395e            ; symbol stub for: strlen
   0x7fff97a33cb7:  48 89 85 50 fc ff ff  mov    qword ptr [rbp - 0x3b0], rax
   0x7fff97a33cbe:  48 89 9d 70 fc ff ff  mov    qword ptr [rbp - 0x390], rbx
   0x7fff97a33cc5:  4c 8d ad 48 fd ff ff  lea    r13, qword ptr [rbp - 0x2b8]

General Purpose Registers:
       rax = 0x00007fff5f400168
       rbx = 0x00007fff5f400528
       rcx = 0x00007fff5f400c40
       rdx = 0x000000010044b9a1  "%s (%lld): %s\n"
       rdi = 0x00007fff7df526b8  __global_locale
       rsi = 0x00007fff7df526b8  __global_locale
       rbp = 0x00007fff5f400410
       rsp = 0x00007fff5f3ffff0
        r8 = 0x000000010044b9a1  "%s (%lld): %s\n"
        r9 = 0x00007fff5f400bf0
       r10 = 0x000000001018f6b4
       r11 = 0x0000000000000451
       r12 = 0x00007fff7df526b8  __global_locale
       r13 = 0x000000010044b9a1  "%s (%lld): %s\n"
       r14 = 0x000000010044b9a1  "%s (%lld): %s\n"
       r15 = 0x0000000000000000
       rip = 0x00007fff97a33c9c  libsystem_c.dylib`__vfprintf + 145
    rflags = 0x0000000000010206
        cs = 0x000000000000002b
        fs = 0x0000000000000000
        gs = 0x0000000000000000


Hash: 3171efa808d0fdded15f3a5471ca9559.46ebc0f49f22e0b2db7e00fe58c1fd1e
ANALYSIS INDICATORS:
--------------------
StopDesc:           EXC_BAD_ACCESS (code=2, address=0x7fff5f3fffe8)
AvNearNull:         False
AvNearSP:           True
BadBeef             False
Access Type:        recursion
Registers:
BlockMov:           False
Weird PC:           False
Weird SP:           False
Suspicious Funcs:
Illegal Insn:       False
Huge Stack:         True

*/

func explode(raw []byte, cmd string) {
	s := `
BUG: Internal error parsing LLDB output!

Something went wrong trying to parse the output of LLDB and we can't continue
without emitting stupid results. If this is a crash that's not worth money,
please open an issue and include the raw LLDB output. If not then just wait, I
guess. :)

LLDB OUTPUT:

`
	panic(fmt.Sprintf("%s %s\nCOMMAND:\n%s\n", s, string(raw), cmd))
}

func mustParseHex(s string, die func()) (n uint64) {
	n, err := strconv.ParseUint(s, 0, 64)
	if err != nil {
		die()
	}
	return
}

func mustAddExtra(prefix string, scanner *bufio.Scanner, ci *crash.Info, die func()) {
	scanner.Scan()
	if !strings.HasPrefix(scanner.Text(), prefix) {
		die()
	}
	ci.Extra = append(ci.Extra, scanner.Text())
}

func mustAdvanceTo(token string, scanner *bufio.Scanner, die func()) {
	for scanner.Scan() {
		if scanner.Text() == token {
			return
		}
	}
	die()
}

func parseIndicators(raw []byte, ci *crash.Info, die func()) {
	// 	ANALYSIS INDICATORS:
	// --------------------
	// StopDesc:           EXC_BAD_ACCESS (code=2, address=0x7fff5f3fffe8)
	// AvNearNull:         False
	// [...]
	scanner := bufio.NewScanner(bytes.NewReader(raw))
	mustAdvanceTo("ANALYSIS INDICATORS:", scanner, die)
	scanner.Scan() // divider row -----
	mustAddExtra("StopDesc:", scanner, ci, die)
	mustAddExtra("AvNearNull:", scanner, ci, die)
	mustAddExtra("AvNearSP:", scanner, ci, die)
	mustAddExtra("BadBeef:", scanner, ci, die)
	mustAddExtra("Access Type:", scanner, ci, die)
	mustAddExtra("Registers:", scanner, ci, die)
	mustAddExtra("BlockMov:", scanner, ci, die)
	mustAddExtra("Weird PC:", scanner, ci, die)
	mustAddExtra("Weird SP:", scanner, ci, die)
	mustAddExtra("Suspicious Funcs:", scanner, ci, die)
	mustAddExtra("Illegal Insn:", scanner, ci, die)
	mustAddExtra("Huge Stack:", scanner, ci, die)
}

func parseExploitable(raw []byte, ci *crash.Info, die func()) {
	scanner := bufio.NewScanner(bytes.NewReader(raw))
	for scanner.Scan() {
		if strings.HasPrefix(scanner.Text(), "Hash:") {
			ff := strings.Fields(scanner.Text())
			if len(ff) < 2 {
				die()
			}
			ci.Hash = ff[1]
			break
		}
	}
}

func parseDisasm(raw []byte, die func()) (crash.Instruction, []crash.Instruction) {
	//Nearby code:
	// libsystem_c.dylib`__vfprintf + 145:
	// -> 0x7fff97a33c9c:  e8 d9 01 04 00        call   0x7fff97a73e7a            ; symbol stub for: localeconv_l
	//    0x7fff97a33ca1:  48 8b 18              mov    rbx, qword ptr [rax]
	//    0x7fff97a33ca4:  80 7b 01 00           cmp    byte ptr [rbx + 0x1], 0x0
	//    0x7fff97a33ca8:  b8 01 00 00 00        mov    eax, 0x1
	//    0x7fff97a33cad:  74 08                 je     0x7fff97a33cb7            ; __vfprintf + 172
	//    0x7fff97a33caf:  48 89 df              mov    rdi, rbx
	//    0x7fff97a33cb2:  e8 a7 fc 03 00        call   0x7fff97a7395e            ; symbol stub for: strlen
	//    0x7fff97a33cb7:  48 89 85 50 fc ff ff  mov    qword ptr [rbp - 0x3b0], rax
	//    0x7fff97a33cbe:  48 89 9d 70 fc ff ff  mov    qword ptr [rbp - 0x390], rbx
	//
	// Stack trace:

	disasm := []crash.Instruction{}
	fault := crash.Instruction{}
	scanner := bufio.NewScanner(bytes.NewReader(raw))
	mustAdvanceTo("Nearby code:", scanner, die)

	// Parse the nearby code
	for scanner.Scan() {

		if strings.HasPrefix(scanner.Text(), "Stack trace:") {
			break
		}

		ff := strings.Fields(scanner.Text())
		// sometimes we get extra crap like:
		// libsystem_kernel.dylib`__pthread_kill:
		if len(ff) == 0 || !((ff[0] == "->") || strings.HasPrefix(ff[0], "0x")) {
			continue
		}

		// The LLDB disasm output with opcodes is nicely formatted. Let's just
		// use theirs.
		colon := strings.Index(scanner.Text(), ":")
		if colon < 0 || len(ff) < 3 {
			die()
		}
		txt := scanner.Text()[colon:]

		if ff[0] == "->" {
			fault = crash.Instruction{
				Address: mustParseHex(strings.TrimSuffix(ff[1], ":"), die),
				Text:    txt,
			}
			disasm = append(disasm, fault)
			continue
		}

		disasm = append(
			disasm,
			crash.Instruction{
				Address: mustParseHex(strings.TrimSuffix(ff[0], ":"), die),
				Text:    txt,
			},
		)

	}
	return fault, disasm
}

func parseRegisters(raw []byte, die func()) (registers []crash.Register) {
	// General Purpose Registers:
	//        rax = 0x0000000000000000
	//        rbx = 0x0000000000000006
	//        rcx = 0x00007fff5fbffc88

	registers = make([]crash.Register, 0, 20)
	scanner := bufio.NewScanner(bytes.NewReader(raw))
	mustAdvanceTo("General Purpose Registers:", scanner, die)

	for scanner.Scan() {
		if scanner.Text() == "" {
			break
		}
		ff := strings.Fields(scanner.Text())
		if len(ff) < 3 {
			die()
		}
		registers = append(
			registers,
			crash.Register{
				Name:  ff[0],
				Value: mustParseHex(ff[2], die),
			},
		)
	}
	return
}

func parseStack(raw []byte, ci *crash.Info, die func()) {
	// Stack trace:
	// * thread #1: tid = 0x9f2fa1f, 0x00007fff97580282 libsystem_kernel.dylib`__pthread_kill + 10, queue = 'com.apple.main-thread', stop reason = signal SIGABRT
	//   * frame #0: 0x00007fff97580282 libsystem_kernel.dylib`__pthread_kill + 10
	//     frame #1: 0x00007fff8e3d54c3 libsystem_pthread.dylib`pthread_kill + 90
	//     frame #2: 0x00007fff97a4cb73 libsystem_c.dylib`abort + 129
	//     frame #3: 0x00000001000b4dc4 pdftoppm`Catalog::getNumPages() [inlined] Object::dictIs(this=0x0000000000000007, dictType=<unavailable>) - 18446744069413843515
	//     frame #4: 0x00000001000029c0 pdftoppm`main(argc=2, argv=<unavailable>) - 18446744069414573631
	//     frame #5: 0x00007fff9a2f35c9 libdyld.dylib`start + 1
	//     frame #6: 0x00007fff9a2f35c9 libdyld.dylib`start + 1

	scanner := bufio.NewScanner(bytes.NewReader(raw))
	mustAdvanceTo("Stack trace:", scanner, die)

	scanner.Scan()
	ff := strings.Fields(scanner.Text())
	if len(ff) < 8 || ff[1] != "thread" {
		die()
	}

	for scanner.Scan() {

		// scan until we run out of frames
		ff := strings.Fields(scanner.Text())
		if len(ff) == 0 {
			break
		}
		// A leading asterisk adjusts all our offsets by 1
		adjust := 0
		if ff[0] == "*" {
			adjust++
		}
		if ff[0+adjust] != "frame" {
			break
		}
		if len(ff) < 3 {
			die()
		}

		var address uint64
		var found bool
		for _, s := range ff {
			if strings.HasPrefix(s, "0x") {
				address = mustParseHex(s, die)
				found = true
			}
		}
		if !found {
			die()
		}
		splits := []string{"???", "???"}
		if len(ff) > 3+adjust {
			// split on backtick (`)
			// libsystem_pthread.dylib`pthread_kill + 90
			rest := strings.Join(ff[3+adjust:], " ")
			splits = strings.SplitN(rest, "`", 2)
			if len(splits) < 2 {
				die()
			}
		}
		entry := crash.StackEntry{
			Address: address,
			Module:  splits[0],
			Symbol:  splits[1],
		}
		ci.Stack = append(ci.Stack, entry)

	}

	// Set the fault to the first frame that's not blacklisted
walk:
	for _, frame := range ci.Stack {
		for _, r := range blacklist {
			if r.MatchString(frame.Module) {
				continue walk
			}
		}
		ci.FaultingFrame = frame
		break
	}

	// don't be too fussy about not finding a stack, here, some crashes set
	// those registers to values that are unreadable as addresses.
	return
}

func parse(raw []byte, cmd string) crash.Info {

	// this just prettifies the rest of the parsers slightly
	die := func() {
		explode(raw, cmd)
	}

	ci := crash.Info{}

	ci.Registers = parseRegisters(raw, die)
	parseStack(raw, &ci, die)
	parseExploitable(raw, &ci, die)
	ci.FaultingInsn, ci.Disassembly = parseDisasm(raw, die)
	parseIndicators(raw, &ci, die)

	return ci
}

func (e *Engine) Run(command []string) (crash.Info, error) {

	pkg, err := build.Import("github.com/bnagy/crashwalk", ".", build.FindOnly)
	if err != nil {
		return crash.Info{}, fmt.Errorf("Couldn't find import path: %s", err)
	}
	tool := filepath.Join(pkg.Dir, "lldb/exploitaben/exploitaben.py")

	cmdSlice := append([]string{tool, "--"}, command...)
	cmdStr := strings.Join(cmdSlice, " ")
	cmd := exec.Command(cmdSlice[0], cmdSlice[1:]...)
	stdout, err := cmd.StdoutPipe()
	if err != nil {
		return crash.Info{}, fmt.Errorf("Error creating pipe: %s", err)
	}
	if err := cmd.Start(); err != nil {
		return crash.Info{}, fmt.Errorf("Error launching tool: %s", err)
	}

	// We don't care about this error because we don't care about exploitaben's exit
	// status.
	out, _ := ioutil.ReadAll(stdout)
	cmd.Wait()

	if bytes.Contains(out, []byte("exited with status")) ||
		bytes.Contains(out, []byte("killing the process...")) {
		// No crash.
		return crash.Info{}, fmt.Errorf("No lldb output for %s", cmdStr)
	}

	ci := parse(out, cmdStr)
	return ci, nil

}
